agent/wsclient/client_test.go

```diff
@@ -0,0 +1,51 @@
+// Copyright 2014-2016 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package wsclient
+
+import (
+	"github.com/aws/amazon-ecs-agent/agent/acs/model/ecsacs"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient/mock/utils"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/credentials"
+	"testing"
+)
+
+// Concurrent writes will cause a panic in the websocket library
+// See https://godoc.org/github.com/gorilla/websocket#hdr-Concurrency
+func TestConcurrentWritesDontPanic(t *testing.T) {
+
+	mockServer, _, _, _, _ := mockwsutils.StartMockServer(t, make(chan bool))
+	defer mockServer.Close()
+
+	req := ecsacs.AckRequest{Cluster: aws.String("test"), ContainerInstance: aws.String("test"), MessageId: aws.String("test")}
+	types := []interface{}{ecsacs.AckRequest{}}
+
+	cs := ClientServerImpl{
+		URL:                mockServer.URL,
+		Region:             "us-east-1",
+		AcceptInvalidCert:  true,
+		CredentialProvider: credentials.AnonymousCredentials,
+		TypeDecoder:        BuildTypeDecoder(types),
+	}
+
+	cs.Connect()
+
+	executeTenRequests := func() {
+		for i := 0; i < 10; i++ {
+			cs.MakeRequest(&req)
+		}
+	}
+	go executeTenRequests()
+	executeTenRequests()
```

Comment:
I like the idea of verifying that all the writes complete. I'll make that change.

Also, would it be reasonable to recover from the panic and fail the test explicitly? Not sure what the best practice is here.